packages/teams-components/src/components/Toolbar/Toolbar.tsx

```diff
@@ -0,0 +1,168 @@
+
+  React.useLayoutEffect(() => {
+    if (!elRef.current?.ownerDocument.defaultView) {
+      return;
+    }
+
+    if (process.env.NODE_ENV !== 'production') {
+      validateToolbarItems(elRef.current);
+    }
+
+    const treeWalker = new HTMLElementWalker(elRef.current, (el) => {
+      if (isAllowedToolbarItem(el) || el === treeWalker.root) {
+        return NodeFilter.FILTER_ACCEPT;
+      }
+
+      return NodeFilter.FILTER_REJECT;
+    });
+
+    reaclcToolbarSpacing(treeWalker);
+
+    const mutationObserver =
+      new elRef.current.ownerDocument.defaultView.MutationObserver(() => {
+        if (!elRef.current) {
+          return;
+        }
+
+        if (process.env.NODE_ENV !== 'production') {
+          validateToolbarItems(elRef.current);
+        }
+
+        // TODO can optimize by only doing recalc of affected elements
+        reaclcToolbarSpacing(treeWalker);
+      });
+
+    mutationObserver.observe(elRef.current, {
+      childList: true,
+    });
+
+    return () => mutationObserver.disconnect();
+  }, []);
+
+  return elRef;
+};
+
+const reaclcToolbarSpacing = (treeWalker: HTMLElementWalker) => {
+  treeWalker.currentElement = treeWalker.root;
+  let current = treeWalker.firstChild();
+  while (current) {
+    recalcToolbarItemSpacing(current, treeWalker);
+
+    treeWalker.currentElement = current;
+    current = treeWalker.nextElement();
+  }
+};
+
+const isAllowedToolbarItem = (el: HTMLElement) => {
+  return (
+    el.classList.contains(toolbarButtonClassNames.root) ||
+    el.classList.contains(toolbarDividerClassNames.root) ||
+    el.classList.contains(toolbarMenuButtonClassNames.root) ||
+    el.classList.contains(toolbarToggleButtonClassNames.root)
+  );
+};
+
+const isPortalSpan = (el: HTMLElement) => {
+  return el.tagName === 'SPAN' && el.hasAttribute('hidden');
+};
+
+const isTabsterDummy = (el: HTMLElement) => {
+  return el.hasAttribute('data-tabster-dummy');
+};
+
+const validateToolbarItems = (root: HTMLElement) => {
+  const children = root.children;
+  for (const child of children) {
+    // TODO is this even possible?
+    if (!isHTMLElement(child)) {
+      continue;
+    }
+
+    if (
+      !isAllowedToolbarItem(child) &&
+      !isPortalSpan(child) &&
+      !isTabsterDummy(child)
+    ) {
+      throw new Error(
+        '@fluentui-contrib/teams-components::Toolbar::Use Toolbar components from @fluentui-contrib/teams-components package only'
+      );
+    }
+  }
+};
+
+const recalcToolbarItemSpacing = (
+  el: HTMLElement,
+  treeWalker: HTMLElementWalker
+) => {
+  treeWalker.currentElement = treeWalker.root;
+  if (el === treeWalker.firstChild() || !isAllowedToolbarItem(el)) {
+    return;
+  }
+
+  if (el.classList.contains(toolbarDividerClassNames.root)) {
+    el.style.marginInlineStart = tokens.spacingHorizontalS;
+    return;
+  }
+
+  treeWalker.currentElement = el;
+  const prev = treeWalker.previousElement();
+  if (prev && prev.dataset.appearance !== 'transparent') {
```

Comment:
I don't think that relying on DOM is a good idea. Registration would work better: we don't expose anything to DOM, we are free to store structure of any complexity

Reply:
will try out registration and compare